### DISCUSSION TITLE
chapter_introduction/what_is_dsa/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#10] [Announcements]
# chapter_introduction/what_is_dsa/

Your first book to learn Data Structure And Algorithm.

https://www.hello-algo.com/chapter_introduction/what_is_dsa/

**Comment by fyxs** [comment]
初学算法冒昧说下，文章很适合我这种初学者，感谢大神。看了本章后根据自己大脑的想象，对于最后的总结，私以为数据结构不是积木，数据是积木，数据结构是积木提供的组织形式等信息，比如不同的形状大小，可连接形式等，而算法操纵这些积木达成某个形态或什么样的形态，也依赖于积木提供的这些信息。甚至包括操纵积木的手或其它什么工具，我认为也是数据结构的部分，是提供的操作积木(数据)的工具——基于学习本文概念后的理解

**Reply by krahets** [reply]
感谢回复！您的解释更加完整清晰～ 是否可以请你按照这个理解更新文章内容，参照 [一起创作本书](https://www.hello-algo.com/chapter_preface/contribution/#_2) 中的方法，你将会作为共同撰稿人在致谢中列出。

> 原文提到 “积木” 时，隐含地包括了积木的结构（数据结构）和积木的组成内容（数据）。例如我们提到 “拼接数组” ，这里的数组就是既指数组中的数据、又隐含了其数据结构。

不过还是期待您可以优化下本章，使之更加清晰

**Reply by Wizarder54088** [reply]
大佬66，这个解释确实更符合数据结构这个定义。

**Reply by AleoXZ** [reply]
这个解释很直观，很形象。

**Reply by longxi102** [reply]
同意，我觉得可以把数据想象成物质，数据结构想象成物质的三态变化，把算法想象成利用物质来加工其他东西，输入的物质量一定，加工过程一定，输出结果就一定

**Reply by zhililab** [reply]
总结常见数据结构及其两两比较的优劣：

1. **数组 (Array)**
   - 优点：快速的随机访问（O(1) 时间复杂度）。
   - 缺点：插入和删除操作需要移动元素，效率低（O(n) 时间复杂度）。

2. **链表 (Linked List)**
   - 优点：插入和删除操作高效（O(1) 时间复杂度）。
   - 缺点：随机访问效率低（O(n) 时间复杂度）。

3. **栈 (Stack)**
   - 优点：操作简单，支持 LIFO（后进先出）操作，高效的插入和删除（O(1) 时间复杂度）。
   - 缺点：只能访问栈顶元素，不支持随机访问。

4. **队列 (Queue)**
   - 优点：操作简单，支持 FIFO（先进先出）操作，高效的插入和删除（O(1) 时间复杂度）。
   - 缺点：只能访问队头和队尾元素，不支持随机访问。

5. **哈希表 (Hash Table)**
   - 优点：快速的插入、删除和查找操作（平均 O(1) 时间复杂度）。
   - 缺点：在哈希冲突时可能退化为链表，最坏情况 O(n) 时间复杂度，内存使用效率较低。

6. **二叉搜索树 (Binary Search Tree, BST)**
   - 优点：支持动态集合操作，插入、删除和查找平均 O(log n) 时间复杂度。
   - 缺点：不平衡时最坏情况 O(n) 时间复杂度。

7. **平衡树 (Balanced Tree, 如 AVL 树、红黑树)**
   - 优点：保证插入、删除和查找操作在 O(log n) 时间复杂度内完成。
   - 缺点：实现复杂度高，插入和删除操作可能需要进行多次旋转和调整。

8. **堆 (Heap)**
   - 优点：支持快速的优先级队列操作，如插入和删除最大/最小值操作在 O(log n) 时间复杂度内完成。
   - 缺点：不支持快速的随机访问。

9. **图 (Graph)**
   - 优点：适用于表示复杂的关系和结构，如社交网络、交通路线等。
   - 缺点：需要占用较大的内存空间，算法复杂度高。

### 数据结构两两比较

1. **数组 vs. 链表**
   - 数组：随机访问快（O(1)），插入删除慢（O(n)）。
   - 链表：插入删除快（O(1)），随机访问慢（O(n)）。

2. **栈 vs. 队列**
   - 栈：LIFO 结构，只能操作栈顶。
   - 队列：FIFO 结构，只能操作队头和队尾。

3. **哈希表 vs. 二叉搜索树**
   - 哈希表：插入、删除和查找平均 O(1)，最坏 O(n)。
   - 二叉搜索树：插入、删除和查找平均 O(log n)，最坏 O(n)。

4. **二叉搜索树 vs. 平衡树**
   - 二叉搜索树：实现简单，可能不平衡。
   - 平衡树：实现复杂，但保证 O(log n) 操作时间。

5. **堆 vs. 二叉搜索树**
   - 堆：优先队列操作快（O(log n)），不支持随机访问。
   - 二叉搜索树：支持快速随机访问（O(log n)），适合动态集合操作。

6. **数组 vs. 哈希表**
   - 数组：固定大小，随机访问快（O(1)）。
   - 哈希表：动态大小，插入和删除快（平均 O(1)）。

7. **图 vs. 链表**
   - 图：适合表示复杂关系，占用内存大。
   - 链表：适合简单线性关系，占用内存小。

with ChatGPT

**Comment by niuren930** [comment]
说来惭愧，工作一年半了，但算法还是小白，这两天打开LeetCode准备开始刷题，突然看到大佬的这本书籍，有种及时雨的感觉。以前总是听说二分查找，心里第一反应就是不懂、难，但是阅读完本章听完大佬的解释，真的是秒懂，很感谢，对本书充满了期待，谢谢。我的微信N-00006，有刷题意愿的小伙伴一起监督一起刷，记得备注刷题哈。感谢大佬，待我阅读完本书再来评价，谢谢！

**Reply by krahets** [reply]
感谢鼓励！～

**Comment by KKStudyCoding** [comment]
「数据结构 Data Structure」是在计算机中组织与存储数据的方式。 看了楼上的评论 我赞同。小白来说可能不知道内存是什么？CPU如何操作内存的数据。数据结构是不是可以先简单粗略的理解为，将数据（我们定义的变量啊，Java中的对象那种概念）存放不同的形式（形式有树、图、链表）这种，存放不同的形式是为了方便不同的操作（类似数学定义上的加减乘除，也有类似生物学上的分类，中药格子中的药材归类，按照所需，找，拿，放）。具体形式长什么样，后续应该有介绍。我也想在后续的阅读过程中添加自己的类似生活中的理解。先模糊理解，在具体精准理解。谢谢您的付出，谢谢你们的付出。

**Reply by krahets** [reply]
谢谢您，非常好的建议！如果您有兴趣可以点击页面右上角 铅笔图标，直接按照这个想法添加内容。我先把这个想法加到 Todo 里，后续会加。

**Comment by tsunamiGG** [comment]
个人意见，关于算法的定义：算法是在有限时间内解决问题的一组指令或操作步骤。

这里的 “解决问题”，应该是解决特定问题。

**Reply by krahets** [reply]
谢谢优化！可否烦请点击页面右上角编辑按钮，更新内容

**Reply by ghost** [reply]
不太理解，为什么要加 特定 两个字

**Reply by krahets** [reply]
@XiaomeiLi111 它“强调”了算法的一组步骤旨在解决某一特定问题，问题具有明确的定义和范围。

**Comment by xiaoyuqiqi** [comment]
y=f(x)

**Comment by tawwxi** [comment]
大佬 pc端 正文 与目录 字体异常模糊 更换chrome edge不管用  黑夜模式也不管用

**Reply by krahets** [reply]
Hi 我感觉可能是你计算机本地缺少字体，一般不会出现这种情况。你用的是什么系统、什么版本？

**Comment by tawwxi** [comment]
您好&nbsp; 我的系统版本为&nbsp;Windows 11, version 22H2 浏览器字体首选项为微软雅黑&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "krahets/hello-algo"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月17日(星期五) 晚上11:20
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [krahets/hello-algo] chapter_introduction/what_is_dsa/ (Discussion #10)





 
Hi 我感觉可能是你计算机本地缺少字体，一般不会出现这种情况。你用的是什么系统、什么版本？

**Reply by krahets** [reply]
那感觉很奇怪了，按说新系统不应该。
您可以试试给电脑切换一个网络环境，比如用手机4g热点

**Comment by NixusAI** [comment]
终于找到好东西了

**Comment by 1713612859** [comment]
直截了当，说法清晰。

**Comment by cyp-happy** [comment]
这个算法的定义其实我之前一直有误解，比如遗传算法等优化算法，输入相同结果也可能不一样吧

**Reply by krahets** [reply]
你是对的。如果引入了随机性，那么多次运行算法可能得到不同的结果。更准确的说法应为：给定相同的输入，在同样的运行条件下，输出始终相同。

请注意，“同样的运行条件”意味着运行时间、环境、随机种子等所有因素都一致，此时伪随机算法产生的随机数是确定的，算法的解也是确定的。

**Reply by Alafun** [reply]
遗传算法是一种统计学算法， 和计算机有输入输出IO的算法不一样

**Reply by krahets** [reply]
@Alafun 遗传算法也是一种算法，也有输入输出

**Comment by Ltruth** [comment]
在作者提供的类比思路下延申，只是疑惑下面的这些问题都是可计算问题吗？不然输入输出结构更适用于计算机而非人，对于人来说，使用感知和创造会更好吧。
数据结构与算法的关系可以用 **积木与说明书 ，建筑材料与蓝图，食材食物和烹饪方法**来形象类比。

每一块不同的积木、每一种建材、每一种食物都有特定的组织和存储方式。比如，从数据结构的角度看，肉类应该冷冻，但香蕉就不适合放冰箱。从算法的角度看，它决定了食物的处理和烹饪方式，榴莲一般不榨汁，香蕉榴莲要先去皮，西瓜要先切开一般不去皮；无硬皮果蔬可以生吃凉拌，肉类就不适合；豆腐脑适合加糖。

最终解决问题还要有操控者，小朋友们搭积木，工程师和建筑工人搭建筑，家庭煮妇或煮夫搭建美食，负责处理计算、具体操作和按照给定的算法或心中的算法控制流程等事项。

**Reply by HM030205** [reply]
豆腐脑要加卤汁

**Reply by samabl** [reply]
可恶的甜党😠

**Reply by imxiaomin** [reply]
加深理解

**Comment by AresInGit** [comment]
四五年前要是有本书，就不至于学的时候这么痛苦了。给K神及小伙伴们点赞！

**Comment by Nexzk** [comment]
很棒！😊

**Comment by coderedeng** [comment]
通俗易懂，非常感谢！

**Comment by zhenxingliu** [comment]
这个搭积木的例子形象生动,比如玩乐高,一堆小零件(数据结构),没有说明书(算法步骤),要想组装起来那几乎不可能.

**Comment by yukiyukixing** [comment]
看到作者积木的比喻，我想到了我经常玩的魔方：

**Reply by yukiyukixing** [reply]
输入数据：乱序的魔方
数据结构：魔方的各个小块以及它们之间的相对位置和关系
算法：旋转魔方的一系列操作步骤，以达到目标状态（这个步骤不是唯一的，可以有很多种，其实也可以认为是不同的公式，对应算法就是不同的算法）
输出数据：解好的魔方

魔方的结构决定了哪些旋转是可能的，而算法决定了如何进行哪些旋转步骤来解好魔方。

**Reply by yukiyukixing** [reply]
使用最短的步骤还原魔方就是时间复杂度最低的算法，但是可能占用你大脑的cpu就多了

**Comment by xtzyyyyy** [comment]
留迹。

**Comment by chickenbreakful** [comment]
小时候扭魔法不知道有说明书这种东西，还以为大家都是自学成才的，后来知道是有公式，但是自己看了还是不懂，可能是不能理解算法深层的含义吧

**Reply by HaroMeng** [reply]
我小时候玩魔方用了大概很长很长时间把一个模仿给复原了，在此之前是不知道有公式这个东西的，完全是自己摸索。后来看了视频教程明白了原理，快了许多，有一个好的引导要远比自己摸索强得多

**Comment by imxiaomin** [comment]
手多动,忆长存

**Reply by krahets** [reply]
满嘴顺口溜 🤣

**Comment by CodingGorit** [comment]
数据结构 和 算法密不可分！
- 数据结构提供了结构化存储数据的方案，以及操作数据的方法
- 算法用来配合数据结构解决特定的问题

**Comment by newcakes** [comment]
打卡第一天：
我啥也不懂！

**Comment by YiXiaoBaiLP** [comment]
这些图自己画的吗？这么好看

**Reply by codezedwow** [reply]
朋友,现在是AI时代

**Comment by Harbour-z** [comment]
* 数据结构为算法提供了结构化存储的数据，以及用于操作数据的方法。
* 数据结构与算法是独立于编程语言的

**Comment by Miles4364** [comment]
如果用做菜去理解，输入数据是原材料（蔬菜），数据结构是原材料的处理方式（切刨剁砍），算法是烹饪方式（蒸煎炒炸），输出数据是成品菜。不知道这样理解对不对？

**Reply by Guanlen** [reply]
我觉得“切刨剁砍”和“蒸煎炒炸”这种动手类的操作都应该属于烹饪方式，也就是属于算法范畴内的；至于这里的数据结构，我认为应该是所要制作的菜品名称及该菜品所包含的各类蔬菜组合和制作方法。比如西红柿炒鸡蛋，原材料（输入数据）是西红柿和鸡蛋，炒制方法（数据结构）给你列出西红柿需要切块和鸡蛋需要打碎后放到容器内搅拌均匀然后倒入锅中翻炒，具体操作步骤（算法）就是按炒制方法去动手炒菜，最后把炒好的西红柿炒鸡蛋盛到盘子里用来食用，这一步也就是数据的输出。

**Comment by levixzheng** [comment]
可以这样理解吗？
拿数学来举例，数据可以看作是题目给出的各种未知数；而数据结构可以看作是数字的各种排列表达方式，比如小数，分数，三角函数甚至是复数；而算法就是求解题目的步骤，来计算出正确的答案。不同的算法使用对应的格式（数据结构）来计算。比如计算角度的**算法**最好用三角函数（**数据结构**）这种表示方法效率更高效。

**Comment by deusyu** [comment]
“数据结构设计是一个充满权衡的过程”，从数据结构设计到软件的架构设计，再到人生，都是trade-off的过程。

**Reply by krahets** [reply]
可谓"有得必有失"

**Comment by USTBliWei** [comment]
清晰易懂

**Comment by lgphone** [comment]
配图非常精美

**Reply by Grit550** [reply]
mid提供技术支持

**Comment by yusui-shun** [comment]
真不错有这种书

**Comment by liangguitao** [comment]
DAY 1 ：LGT  自我总结：
算法是离不开数据结构的
就像一个圆柱容器，把他完整的不留缝隙的填满，面前有一桶水，一堆棉花，一堆砖头，填满的过程就是算法，而材料就是数据结构，在具体问题上这两个缺一不可，需要解决问题的人自己权衡利弊。

**Reply by mainiofun** [reply]
算法是一种过程，数据结构是一种组织数据的方式： 算法是一系列定义好的步骤，用于解决特定问题或执行特定任务。数据结构则是在计算机中组织和存储数据的方式。你这个这个比喻虽然生动形象，但存在一些概念上的不准确之处。算法并非填满容器的过程： 在比喻中，将填满容器的过程描述为算法可能引起误解。算法更准确地表示为解决问题的方法或步骤。填满容器更像是数据的组织和存储过程，而不是算法

**Reply by GalaxyLee** [reply]
把编程看作盖房子的话（自己脑补一下建筑师角色），算法你可以想象成自己的设计和规划，它定义了完成特定任务所需的一系列步骤和操作。数据结构可以类比为建筑材料，它们是用于存储和组织数据的方式。
而且，算法可以是通用的，也可以是为特定数据结构设计的。例如，排序算法（如快速排序、归并排序）可以应用于任何类型的数据集合。数据结构为算法提供了执行操作的基础，如数组、链表、栈、队列、树、图等。

**Comment by cloudyuyuyu** [comment]
坚持学下去20240108

**Reply by krahets** [reply]
加油！

**Comment by zhangjunchang** [comment]
这本书确实很适合新手，非常感谢这样免费开源的中国作者。

**Comment by CNiq527** [comment]
> 链表相较于数组，在数据添加和删除操作上更加便捷，但牺牲了数据访问速度。

这个数据访问速度，不够精确，是不是要说明下，“随机访问”?

**Reply by krahets** [reply]
由于缓存局部性不同，数组的顺序访问也比链表快。（当然，随机访问的差别相对更大一些）

**Comment by deyu9999** [comment]
积木的比喻第一次看到，挺好

**Comment by GalaxyLee** [comment]
<strong>数据结构</strong>的定义:<strong><em>数据结构是相互之间存在一种或多种特定关系的数据元素的集合。</em></strong>在任何问题中，数据元素都不是孤立存在的，它们之间存在某种关系，这种数据元素相互之间的关系称为结构（Structure）。
<strong>数据元素</strong>的定义：数据元素是数据的基本单位。
<strong>数据</strong>的定义：数据是信息的载体，是描述客观事物属性的数、字符及所有能输入到计算机中并被计算机程序识别和处理的符号的集合。<strong><em>数据是计算机程序加工的原料。</em></strong>
表 1-1将“输入数据”比喻为“未拼装的积木”，个人拙见，“积木”是否应该解释成“输入数据的数据元素”呢？

**Comment by LotusQing** [comment]
感谢免费分享，好人一生平安

**Comment by junsircoding** [comment]
以我的中文水平和对内容的理解，我觉得 *1.2.3* 小节的开头对 **数据结构** 和 **算法** 之间关系的描述中，第二点应该是 `数据结构是算法发挥作用的舞台。` 。也就是说，数据结构是舞台，算法在数据结构上发挥作用。

不知各位是否同意？

**Reply by krahets** [reply]
Hi，两者都可以看作是对方的舞台。在算法的加持下，数据结构才会“跳动”起来～

**Reply by jackchen-hk** [reply]
可以看作“数据结构”搭台，“算法”唱戏

**Comment by ncufresher** [comment]
应该可以这样理解，数据结构就是对现实世界的一个抽象，用来描述现实世界，而算法就是在数据结构上的操作，就是现实中的解决方法。

**Reply by qwer-der** [reply]
赞

**Comment by NgxuAnGMH** [comment]
任何高级编程语言的官方文档中，都会介绍基本数据类型，这些都是组合成“类”或者“结构体”（引用类型）的基本单元。所以程序的世界，其实就是从基础数据类型（int/float/bool/char/byte..）等等一步步向上抽象的。

也因此，我们理解的业务系统（淘宝京东拼多多/抖音短视频/腾讯微信..）现实中的所有概念，例如说“用户资料”，都是经过组装结合到一起才实现的。

而数据结构就是这些东西“存储的形式”，而“算法”更多的是对该数据结构进行“常见操作”（增删改查）的高效设计。“数据结构与算法”必须搭配工作。

**Comment by duzh** [comment]
应该可以这样看算法和程序开发的关系，把一个项目比作一个个的建筑物，例如写字楼，而在建立大楼前要想好怎么建立用什么建立，这些可以看作算法，使用最高效的方式建立，而开发语言就是具体内容了。。

**Comment by Knowledge-yc** [comment]
这个比喻非常nice

**Comment by niaidaye** [comment]
算法就是，按照问题，来将，特有的数据结构进行，归纳整合，得到最佳合适的输出

**Comment by suhq24** [comment]
数据结构定义中的第三点“提供简洁的数据表示和逻辑信息”，是什么意思呢？

**Reply by tiam-bloom** [reply]
就是表示出数据的信息以及数据之间的逻辑吧

**Reply by precious0612** [reply]
我觉得应该表达的意思是，相对于杂乱无序的数据，数据结构具有简化的数据表示并含有一定内在逻辑。
比如说：3、7、1、0、6、...（后面也是无序的数）与 0、1、2、3、4、... 之间的区别。
自然数不仅有只需要几个数据能表示整个集合（简洁的数据表示），并且它是递增的（逻辑信息）

**Comment by krahets** [comment]
Hi，建议想要在评论区打卡的同学，只在每一章的[封面小节](https://www.hello-algo.com/chapter_introduction/)的评论区中打卡～
（内容小节的评论区主要用于提出见解、问题和建议，打卡评论会使讨论内容分散）

祝阅读愉快！

**Comment by 1262706090** [comment]
我对“包含清晰的输入和输出定义。”和“各步骤都有确定的含义，在相同的输入和运行条件下，输出始终相同。”这个有些疑问。我看了其他介绍，说算法可以没有输入，那此时可不可以有多个输出，如果可以是不是违背了确定性的原则

**Reply by krahets** [reply]
> 算法可以没有输入，那此时可不可以有多个输出

1. 此时我们可以把输入看作是“空”，即空输入也可看作是一种特殊的输入；
2. 如果“运行条件”完全确定，那么输出也应该是确定的；如果多次输出不同，那么通常是算法内部某些环节引入了随机性。

**Reply by 1262706090** [reply]
奥 ，好的。谢谢。

**Comment by KSTLZY** [comment]
已阅

**Reply by chenghao27** [reply]
6

**Comment by Toin1779712664** [comment]
算法具有以下特性：
·准确性：对一个数据进行相应地运算，它的结果是准确的且该算法能满足所需设计要求产生准确的输出结果。
·唯一性：设计一个算法时，如果传入一个数据到这个程序它产生不同的结果那么它的运算结果就不是唯一的。对算法的要求是产生一个唯一且准确的结果。
·有穷性：对传入的数据需要在有效地步骤中进行处理，若该算法不是有穷性则它不是一个可用的算法
·可输入输出：数据的输入和对数据处理后的结果进行输出。
数据结构有以下特性：
·对数据进行有效地性组织：如定义一个Person结构那么需要知道这个Person里具体需要定义哪些数据类型。
·对存放的结构（存储空间结构）有效地选择对应的存储结构：采用线性结构存储还是非线性结构存储。
·数据的存取操作：对数据的增、删、改、查和内存回收等基本操作。

**Comment by jeanleeean** [comment]
自问自答
    算法是什么：算法就是解决问题的一系列步骤，并且要在有限的时间内

    算法有哪些特性：
        具有输入和输出；
        可行性：时间有限、空间有限、操作步骤有限；
        确定性：在相同的条件下，同样的输入对应同样的输出。
    --------------------------------------------------------------------------------------------------

    数据结构是什么：
        本质上就是存储数据的，与算法相结合起来，共同解决问题。为了使效率最大化，在选择数据结构时，往往希望存储尽量少占用内存空间、操作数据方便、数据的表示和逻辑信息清晰简单，但这三者通常很难做到统一，往往需要根据需求来权衡。比如链表比数组更易增加和删除数据，但访问速度更慢。

**Comment by jeanleeean** [comment]
自问自答
算法是什么：算法就是解决问题的一系列步骤，并且要在有限的时间内
--------------------------------------------------------------------------------------------------

算法有哪些特性：
    具有输入和输出；
    可行性：时间有限、空间有限、操作步骤有限；
    确定性：在相同的条件下，同样的输入对应同样的输出。
--------------------------------------------------------------------------------------------------

数据结构是什么：
    本质上就是存储数据的，与算法相结合起来，共同解决问题。为了使效率最大化，在选择数据结构时，往往希望存储尽量少占用内存空间、操作数据方便、数据的表示和逻辑信息清晰简单，但这三者通常很难做到统一，往往需要根据需求来权衡。比如链表比数组更易增加和删除数据，但访问速度更慢。

**Comment by wangming27** [comment]
算法就是方法。

**Comment by Umizhang0910** [comment]
我的理解数据结构就是存储数据的方法（容器、原理），算法就是对数据结构的一系列操作步骤。

**Comment by kyotuni** [comment]
个人感觉数据结构字如其名就像是一个架构，而算法就是通过数据结构来实现问题的解决

**Comment by zaozaozuibang** [comment]
算法就是解决问题的一些步骤，数据结构包括逻辑结构、物理结构（存储数据）以及包括操作数据

**Comment by xuaidongx** [comment]
数据结构：数据的存储方式及操作方法，可携带逻辑关系（如图、树），是算法的基础。
算法：在有限条件下解决问题的一组明确步骤或规则，用于实现特定目标。

**Comment by dashen01** [comment]
程序=数据结构+算法

**Comment by VitalFlare** [comment]
感谢分享，持续阅读。

**Comment by njjnjjnjj** [comment]
写得好啊，后面拼积木的比喻很恰当，易于理解👍👍👍👍👍👍

**Comment by zhishaoni** [comment]
搭积木=秒了！

**Comment by Plum-Yu** [comment]
算法赋予数据结构生命，数据结构赋予算法灵魂。

**Comment by WeChatWeGPT** [comment]
和控制理论、机械传动思想类似

**Comment by HighBigSky** [comment]
大学《操作系统》的课本上知道了 程序 = 算法 + 数据结构

**Comment by Knight-It** [comment]
程序=数据结构+算法

**Comment by Marigold-Lan** [comment]
太棒啦

**Comment by 3218085097** [comment]
在有限时间内通过某些手段解决特定问题。
既要关注抽象的概念，也要考虑实际使用。

**Comment by 2yyue** [comment]
# 菜狗重拾计算机

**Comment by coronabear** [comment]
程序=数据结构+算法

**Comment by Jay20040104** [comment]
# 数据结构是团队中不同的角色，算法是找出解决问题的最佳职能分配

**Comment by AnakinZ** [comment]
“数据结构与算法是独立于编程语言的”
难怪，入门级别的编程语言的书籍都不会特定去提这个的知识点。

**Comment by LittleFatHero** [comment]
数据结构是数据的存储结构，算法是操作数据的方法，数据结构是为算法服务的，算法必须作用在特定的数据结构之上。数据结构是程序的生命，算法是程序的灵魂！

**Comment by IT-YYP** [comment]
是不是可以说算法无处不在，任何形式都可以以算法的形式呈现

**Comment by Jhcnsn** [comment]
奈斯

**Comment by zhaoguhong** [comment]
数据结构是躯体，算法是灵魂！

**Comment by 3336290034** [comment]
Nice！

**Comment by xbsheng** [comment]
程序 = 数据结构 + 算法

**Reply by hongxin01** [reply]
你可以获得图灵奖了

**Comment by Ambrogram** [comment]
数据结构是图书馆的藏书体系，决定了书本如何被分类、索引和存放；算法则是寻书的智慧和路径，让人可以高效地找到所需知识。

**Comment by Ivy129** [comment]
数据结构与算法，很有意思

**Comment by 1aa-1** [comment]
# 程序是身体，算法是灵魂

**Reply by CrescentZM** [reply]
打卡

**Comment by CYX103123** [comment]
##python里面自己写的链表应该不如直接的数组快

**Comment by CYX103123** [comment]
# python里面自己写的链表应该不如直接的数组快

**Comment by zhengads** [comment]
数据结构是程序中数据存在的不同的形式与载体，算法是程序执行任务并产生所需结果的途径与方法

**Reply by zhengads** [reply]
再次启程，加油加油！

**Comment by Sword-Holder-L** [comment]
数据结构决定了数据的“组织形态”，而这个形态直接决定了什么样的“算法”对它来说是高效的。
数据结构是“怎么存”，算法是“怎么干”。
你想“怎么干”（算法），常常取决于你“怎么存”（数据结构）。
你想快速找到一个人？
如果人是按学号排好序的名单（有序数据结构），你可以用对半找（高效算法） 的方法，飞快找到。
如果人是随便坐的（无序数据结构），你只能一个个问（低效算法），慢得很。
你“怎么存”（数据结构），也决定了你“干什么”最方便。
你用的是排队打饭的队伍（队列数据结构），那么“先来的人先打饭”这个事干起来就最方便、最公平。

**Comment by LiLi-0226** [comment]
1.2打卡

**Comment by wink0521** [comment]
什么是数据结构：数据结构定义了存储规则，如物品怎么存放，怎么取。
什么是算法：对数据进行处理的指令
关系：相互依赖，互相影响。

**Comment by Zay-074** [comment]


**Reply by Hellwrod** [reply]
你是在做公务员行测考试吗兄弟。

**Comment by Xiaoha3-3** [comment]
理解是：数据结构是数据被存放的形态
算法是处理这些数据的，依照数据结构来设计对应的算法来处理数据？

**Comment by Ke-Albert** [comment]
算法和数据结构密不可分。一些工程性问题，首先需要设计挑选合适的数据结构去存储数据，然后根据这个数据结构去设计高效的算法进行处理。每一个数据结构的存在都是有它当初被设计出来的理由的。可以多讲讲每一个数据结构被设计出来的背景小故事之类的

**Comment by xmzd-S** [comment]
算法是解决问题的，解决问题就有输入和输出，单单从输入和输出来看，算法的作用就是将输入的数据“映射”为输出的数据。
而数据结构是为了更方便的映射而对数据进行的一种组织形式。

**Comment by FORNCY** [comment]
开始学习第一天！

**Comment by xy0707** [comment]
开始学习，加油！

**Comment by lichuanjie31-prog** [comment]
**编程语言是解决问题的一个手段，重要的是如何解决这个问题，而我们恰恰需要学习的就是后者**
